### ISSUE TITLE
[Bug]: GET /api/list-files HTTP/1.1" 401 Unauthorized     

### ISSUE THREAD

**Issue by lambdaofgod** [All-Hands-AI/OpenHands#5574]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Describe the bug and reproduction steps

When I run the command from README[^1] I can open the UI, use the assistant, but I can't access the files it generated. In the logs I get

```
openhands:WARNING: middleware.py:147 - Missing Authorization header                                                                                                                                     
INFO:     172.17.0.1:38972 - "GET /api/list-files HTTP/1.1" 401 Unauthorized     
```

[^1] the only difference is that I changed the port to 3001

### OpenHands Installation

Docker command in README

### OpenHands Version

docker.all-hands.dev/all-hands-ai/openhands:0.15

### Operating System

Linux

### Logs, Errors, Screenshots, and Additional Context

![Screenshot 2024-12-12 at 22-29-00 OpenHands](https://github.com/user-attachments/assets/4e0aaf1c-39aa-4ea9-96a6-8a766a4d294b)

**Comment by mamoodi** [comment]
It looks like you aren't running into this issue on main?
If that's the case, it's likely a bug already fixed and we just need to release it. Although I don't see a specific fix for this in main.

If you specify 0.15.2 as the image instead of 0.15 for both the OpenHands image and the nikolaik image does it work?

**Comment by enyst** [comment]
@mamoodi FWIW I'm hitting this quite a bit today on app.all-hands.dev. Is it supposed to be updated?

**Comment by mamoodi** [comment]
@enyst right now app.all-hands.dev is running 0.15.2. So if this was fixed in main, it's not out there yet.

**Comment by mamoodi** [comment]
Seeing this a fair amount now. CC @amanape when you're able to do a quick check 🙏

**Comment by autonomousit** [comment]
I see the same thing when running in development mode and using `make run` (on `main` but checked out the `0.16.1` release tag). Switching the frontend and backend ports around in `Makefile` so that the backend port == 3001 and frontend port == 3000 resolves the issue.

**Comment by S0ly** [comment]
I also have this error on the official docker

**Comment by autonomousit** [comment]
So when I see these symptoms in 0.16 they are tied to a different part of the code (original report for 0.15 had `middleware.py:147 -  Missing Authorization header`). I don't think the original issue is the same as mine although the end result is the same. Happy to move this to a separate issue if necessary but it may also not be a difficult fix if I understand it correctly.

I still get a `"GET /api/list-files HTTP/1.1" 401 Unauthorized` error and the end result is the workspace files fail to load but the errors related to it are:
- `auth.py:27 Invalid Token`
- `middleware.py:137` Invalid Token`

Both of these errors are originating from the fact that `get_sid_from_token()` in `auth.py` is failing to decode the JWT token properly. In my case I had just switched from using the default Docker deployment to using Development Mode.

I haven't managed to make my way through the code base enough to work out whether the JWT token for the backend server is being passed by my browser or by the frontend server but is it possible this is just due to re-use of JWT tokens between instances of OpenHands when the JWT secret is randomly set by `get_or_create_jwt_secret()` in `utils.py`? In Development Mode you obviously prevent this by setting the JWT secret in `config.toml` but I'm not sure how you would avoid this when using the default Docker deployment which isn't persistent except for using a private mode tab or clearing cache between re-deployments of the container. @S0ly appears to be in this situation.

**Comment by mamoodi** [comment]
There's been some changes to the persistent of those settings on main. Are you able to give the main image a shot and see if you still get these errors? @autonomousit 

```
docker run -it --rm --pull=always \
    -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:main-nikolaik \
    -e LOG_ALL_EVENTS=true \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -v ~/.openhands-state:/.openhands-state \
    -p 3000:3000 \
    --add-host host.docker.internal:host-gateway \
    --name openhands-app \
    docker.all-hands.dev/all-hands-ai/openhands:main
```

**Comment by autonomousit** [comment]
@mamoodi - so a couple of issues in trying to test this for you.

**TLDR**: I find it hard to replicate this issue, even on the `0.16.1` commit, it may be better for someone else to try this who is seeing it consistently. I also can't get the latest 2 commits on `main` to initialise the app correctly which prevents testing this issue.

- This issue already appeared to resolve itself for me when using Dev Mode on the `0.16.1` commit simply by going to bed and coming back the next day so hard to replicate it in testing. Given that plus the fact that changing the frontend port number to be different to the Docker deployment port fixed the issue for me and that the implementation seems port agnostic was what made me think this was something being cached in the browser.
- When deploying off `main` as you asked I get timeouts waiting for the client - the `/alive` healthcheck to client never returns successfully until it hits the max retries/timeout.
- When deploying `main` using Dev Mode I just get a blank white screen in the browser.
- Commits `5872` and `5874` on `main` are the specific commits I was testing with

**Comment by mamoodi** [comment]
Yeah there's a lot of moving parts right now on main but not great that you can't run it...
Probably some issues still need to be ironed out.
Thanks for trying this out. Really appreciate it. 🙏 

This might be resolved on 0.18 but will keep my eyes open.

**Comment by autonomousit** [comment]
@mamoodi - I think something is messed up in my environment. I tried reverting to the `0.16.1` in dev mode and the agent never becomes ready, always initialising. I then tried the 0.16 docker image which was previously working for me and I just get "Agent has stopped" status forever. Is there any hidden state or other temp stuff I should be looking for and cleaning out to make sure my environment is reset to a blank state? My expectation would be that the docker run method would be fully self-contained with no state carried over (I also cleared my browser cache). At this point I can't run OpenHands using either docker or dev mode on multiple versions so it definitely feels like an environment issue.

Here's the exact docker run command I used:

```bash
docker run -it --rm --pull=always \
    -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:0.16-nikolaik \
    -e LOG_ALL_EVENTS=true \
    -v /var/run/docker.sock:/var/run/docker.sock \
    -p 3000:3000 \
    --add-host host.docker.internal:host-gateway \
    --name openhands-app \
    docker.all-hands.dev/all-hands-ai/openhands:0.16
 ```

**Comment by enyst** [comment]
I'm not sure what's wrong, @autonomousit , just a few quick thoughts:

For development mode: https://github.com/All-Hands-AI/OpenHands/blob/main/Development.md
`setup-config` step is not necessary, can skip of course.

To start in various ways, also: https://docs.all-hands.dev/modules/usage/how-to/debugging

And: can you add in config.toml these: https://docs.all-hands.dev/modules/usage/how-to/persist-session-data
With your own location/names, _new_ values set by you now. Just in case the problem has to do with what we do in case they're not set. (feel free to remove/clean up after, if you want)

**Comment by autonomousit** [comment]
Thanks @enyst I will try and report back. Are you guys developing locally or inside a container? If in a container I assume it's the same nikolaik family of images as used for deployment?

**Comment by enyst** [comment]
Yes, a container but only for the runtime, not the "outside" app docker container. No setting whatsoever works to download it, and it's building on it as necessary.

**Comment by autonomousit** [comment]
Thanks @enyst - adding the persistent storage settings did resolve the issue when running in dev mode, still having some issues when deploying from docker. All of this seems to be coming down to environment setup and on-boarding issues and I thought I might be able to help with this more systematically instead of via all these various issues. I opened a feature request #5878 to discuss it further.

**Comment by mamoodi** [comment]
I think currently there is an issue swapping between different versions, especially when `main` is involved due to the settings migration.

**Comment by mamoodi** [comment]
@lambdaofgod are you seeing this in 0.21.0. There's been a fair amount of changes around this and seems to work fine from what I've tried on different systems.
